File: gestionnaire_mdp.py
import base64
import json
import os
import secrets
import string
import sys
from getpass import getpass
from pathlib import Path

try:
    from cryptography.fernet import Fernet, InvalidToken
    from cryptography.hazmat.primitives import hashes
    from cryptography.hazmat.primitives.kdf.pbkdf2 import PBKDF2HMAC
except ImportError:
    print("La bibliotheque 'cryptography' est manquante.")
    print("Installez-la avec : py -m pip install cryptography")
    sys.exit(1)


FICHIER_COFFRE = Path(__file__).with_name("coffre_mdp.json")
VERSION_COFFRE = 1
ITERATIONS_PBKDF2 = 600000


def afficher_titre(titre):
    largeur = 72
    print()
    print("=" * largeur)
    print(titre.center(largeur))
    print("=" * largeur)


def pause():
    input("\nAppuyez sur Entree pour continuer...")


def lire_texte(message):
    return input(message).strip()


def lire_texte_obligatoire(message):
    valeur = ""
    while valeur == "":
        valeur = lire_texte(message)
        if valeur == "":
            print("Cette valeur ne peut pas etre vide.")
    return valeur


def lire_secret(message):
    if "SPYDER_KERNEL_ID" in os.environ:
        print("Info Spyder : le mot de passe sera visible dans cette console.")
        return input(message)

    try:
        return getpass(message)
    except Exception:
        return input(message)


def demander_oui_non(message):
    while True:
        reponse = lire_texte(message + " (o/n) ").lower()
        if reponse in ("o", "oui"):
            return True
        if reponse in ("n", "non"):
            return False
        print("Reponse attendue : o ou n.")


def deriver_cle(mot_de_passe_maitre, sel, iterations):
    kdf = PBKDF2HMAC(
        algorithm=hashes.SHA256(),
        length=32,
        salt=sel,
        iterations=iterations,
    )
    cle = kdf.derive(mot_de_passe_maitre.encode("utf-8"))
    return base64.urlsafe_b64encode(cle)


def chiffrer_donnees(fernet, entrees):
    donnees_json = json.dumps(entrees, ensure_ascii=False, indent=2)
    return fernet.encrypt(donnees_json.encode("utf-8")).decode("utf-8")


def dechiffrer_donnees(fernet, texte_chiffre):
    donnees = fernet.decrypt(texte_chiffre.encode("utf-8"))
    return json.loads(donnees.decode("utf-8"))


def sauvegarder_coffre(coffre, fernet, entrees):
    coffre["donnees"] = chiffrer_donnees(fernet, entrees)
    with open(FICHIER_COFFRE, "w", encoding="utf-8") as fichier:
        json.dump(coffre, fichier, ensure_ascii=False, indent=2)


def verifier_force_mdp(mot_de_passe):
    erreurs = []

    if len(mot_de_passe) < 10:
        erreurs.append("- au moins 10 caracteres")
    if not any(caractere.islower() for caractere in mot_de_passe):
        erreurs.append("- une lettre minuscule")
    if not any(caractere.isupper() for caractere in mot_de_passe):
        erreurs.append("- une lettre majuscule")
    if not any(caractere.isdigit() for caractere in mot_de_passe):
        erreurs.append("- un chiffre")
    if not any(caractere in "!@#$%&*()-_=+?" for caractere in mot_de_passe):
        erreurs.append("- un caractere special parmi !@#$%&*()-_=+?")

    return erreurs


def creer_mot_de_passe_maitre():
    while True:
        print("\nCreation du mot de passe maitre.")
        print("Conseil : utilisez une phrase longue que vous pouvez retenir.")
        mot_de_passe = lire_secret("Nouveau mot de passe maitre : ")
        confirmation = lire_secret("Confirmez le mot de passe maitre : ")

        if mot_de_passe != confirmation:
            print("Les deux mots de passe ne correspondent pas.")
            continue

        erreurs = verifier_force_mdp(mot_de_passe)
        if erreurs:
            print("\nLe mot de passe maitre est faible. Il manque :")
            for erreur in erreurs:
                print(erreur)
            if not demander_oui_non("Voulez-vous quand meme l'utiliser"):
                continue

        return mot_de_passe


def creer_coffre():
    afficher_titre("CREATION DU COFFRE-FORT")
    print("Aucun coffre n'a ete trouve.")
    print("Le programme va creer un coffre local chiffre.")

    mot_de_passe_maitre = creer_mot_de_passe_maitre()
    sel = secrets.token_bytes(16)
    cle = deriver_cle(mot_de_passe_maitre, sel, ITERATIONS_PBKDF2)
    fernet = Fernet(cle)

    coffre = {
        "version": VERSION_COFFRE,
        "algorithme": "Fernet + PBKDF2-HMAC-SHA256",
        "iterations": ITERATIONS_PBKDF2,
        "sel": base64.b64encode(sel).decode("utf-8"),
        "verificateur": fernet.encrypt(b"mot_de_passe_correct").decode("utf-8"),
        "donnees": "",
    }

    entrees = []
    sauvegarder_coffre(coffre, fernet, entrees)
    print("\nCoffre cree avec succes.")
    return coffre, fernet, entrees


def charger_fichier_coffre():
    with open(FICHIER_COFFRE, "r", encoding="utf-8") as fichier:
        return json.load(fichier)


def ouvrir_coffre_existant():
    afficher_titre("OUVERTURE DU COFFRE-FORT")
    coffre = charger_fichier_coffre()
    sel = base64.b64decode(coffre["sel"])
    iterations = int(coffre.get("iterations", ITERATIONS_PBKDF2))

    for tentative in range(1, 4):
        mot_de_passe_maitre = lire_secret("Mot de passe maitre : ")
        cle = deriver_cle(mot_de_passe_maitre, sel, iterations)
        fernet = Fernet(cle)

        try:
            fernet.decrypt(coffre["verificateur"].encode("utf-8"))
            entrees = dechiffrer_donnees(fernet, coffre["donnees"])
            print("\nCoffre ouvert avec succes.")
            return coffre, fernet, entrees
        except InvalidToken:
            print("Mot de passe maitre incorrect.")
            print("Tentative", tentative, "sur 3.")

    print("\nAcces refuse. Le coffre reste chiffre.")
    sys.exit(1)


def ouvrir_ou_creer_coffre():
    if FICHIER_COFFRE.exists():
        return ouvrir_coffre_existant()
    return creer_coffre()


def generer_mot_de_passe(longueur):
    minuscules = string.ascii_lowercase
    majuscules = string.ascii_uppercase
    chiffres = string.digits
    speciaux = "!@#$%&*()-_=+?"
    alphabet = minuscules + majuscules + chiffres + speciaux

    longueur = max(longueur, 8)
    caracteres = [
        secrets.choice(minuscules),
        secrets.choice(majuscules),
        secrets.choice(chiffres),
        secrets.choice(speciaux),
    ]

    while len(caracteres) < longueur:
        caracteres.append(secrets.choice(alphabet))

    melangeur = secrets.SystemRandom()
    melangeur.shuffle(caracteres)
    return "".join(caracteres)


def afficher_resume_entree(entree, numero):
    print(str(numero) + ".", entree["service"], "-", entree["identifiant"])


def trouver_entrees(entrees, terme):
    terme = terme.lower()
    resultats = []

    for index, entree in enumerate(entrees):
        service = entree["service"].lower()
        identifiant = entree["identifiant"].lower()
        if terme in service or terme in identifiant:
            resultats.append(index)

    return resultats


def choisir_entree(entrees, message):
    if not entrees:
        print("Le coffre ne contient aucune entree.")
        return None

    terme = lire_texte_obligatoire(message)
    resultats = trouver_entrees(entrees, terme)

    if not resultats:
        print("Aucun resultat trouve.")
        return None

    print("\nResultats :")
    for position, index in enumerate(resultats, start=1):
        afficher_resume_entree(entrees[index], position)

    while True:
        choix = lire_texte("Numero de l'entree a selectionner : ")
        if choix.isdigit():
            position = int(choix)
            if 1 <= position <= len(resultats):
                return resultats[position - 1]
        print("Choix invalide.")


def ajouter_entree(entrees):
    afficher_titre("AJOUTER UN MOT DE PASSE")
    service = lire_texte_obligatoire("Service ou site : ")
    identifiant = lire_texte_obligatoire("Identifiant ou email : ")

    if demander_oui_non("Voulez-vous generer un mot de passe fort"):
        longueur_texte = lire_texte("Longueur souhaitee (16 par defaut) : ")
        longueur = 16
        if longueur_texte.isdigit():
            longueur = int(longueur_texte)
        mot_de_passe = generer_mot_de_passe(longueur)
        print("Mot de passe genere :", mot_de_passe)
    else:
        mot_de_passe = lire_secret("Mot de passe a enregistrer : ")
        while mot_de_passe == "":
            print("Le mot de passe ne peut pas etre vide.")
            mot_de_passe = lire_secret("Mot de passe a enregistrer : ")

    note = lire_texte("Note facultative : ")

    entrees.append(
        {
            "service": service,
            "identifiant": identifiant,
            "mot_de_passe": mot_de_passe,
            "note": note,
        }
    )

    print("\nEntree ajoutee.")


def lister_services(entrees):
    afficher_titre("LISTE DES SERVICES")

    if not entrees:
        print("Le coffre est vide.")
        return

    for index, entree in enumerate(entrees, start=1):
        afficher_resume_entree(entree, index)


def rechercher_entree(entrees):
    afficher_titre("RECHERCHER UNE ENTREE")
    index = choisir_entree(entrees, "Recherche par service ou identifiant : ")
    if index is None:
        return

    entree = entrees[index]
    print("\nService      :", entree["service"])
    print("Identifiant  :", entree["identifiant"])
    print("Note         :", entree["note"] if entree["note"] else "Aucune")

    if demander_oui_non("Afficher le mot de passe"):
        print("Mot de passe :", entree["mot_de_passe"])
    else:
        print("Mot de passe : ********")


def modifier_entree(entrees):
    afficher_titre("MODIFIER UNE ENTREE")
    index = choisir_entree(entrees, "Entree a modifier : ")
    if index is None:
        return False

    entree = entrees[index]
    print("\nLaissez vide pour conserver la valeur actuelle.")

    nouveau_service = lire_texte("Nouveau service [" + entree["service"] + "] : ")
    nouvel_identifiant = lire_texte("Nouvel identifiant [" + entree["identifiant"] + "] : ")
    nouvelle_note = lire_texte("Nouvelle note [" + entree["note"] + "] : ")

    if nouveau_service:
        entree["service"] = nouveau_service
    if nouvel_identifiant:
        entree["identifiant"] = nouvel_identifiant
    if nouvelle_note:
        entree["note"] = nouvelle_note

    if demander_oui_non("Changer le mot de passe"):
        if demander_oui_non("Generer un nouveau mot de passe fort"):
            longueur_texte = lire_texte("Longueur souhaitee (16 par defaut) : ")
            longueur = 16
            if longueur_texte.isdigit():
                longueur = int(longueur_texte)
            entree["mot_de_passe"] = generer_mot_de_passe(longueur)
            print("Nouveau mot de passe :", entree["mot_de_passe"])
        else:
            nouveau_mdp = lire_secret("Nouveau mot de passe : ")
            if nouveau_mdp:
                entree["mot_de_passe"] = nouveau_mdp

    print("\nEntree modifiee.")
    return True


def supprimer_entree(entrees):
    afficher_titre("SUPPRIMER UNE ENTREE")
    index = choisir_entree(entrees, "Entree a supprimer : ")
    if index is None:
        return False

    entree = entrees[index]
    print("\nSelection :", entree["service"], "-", entree["identifiant"])
    if demander_oui_non("Confirmer la suppression"):
        del entrees[index]
        print("Entree supprimee.")
        return True

    print("Suppression annulee.")
    return False


def outil_generation():
    afficher_titre("GENERATEUR DE MOTS DE PASSE")
    longueur_texte = lire_texte("Longueur souhaitee (16 par defaut) : ")
    longueur = 16

    if longueur_texte.isdigit():
        longueur = int(longueur_texte)

    mot_de_passe = generer_mot_de_passe(longueur)
    print("\nMot de passe genere :", mot_de_passe)
    print("Vous pouvez l'utiliser lors de l'ajout d'une entree.")


def afficher_menu():
    afficher_titre("GESTIONNAIRE DE MOTS DE PASSE")
    print("1. Ajouter un mot de passe")
    print("2. Rechercher une entree")
    print("3. Lister les services")
    print("4. Modifier une entree")
    print("5. Supprimer une entree")
    print("6. Generer un mot de passe fort")
    print("7. Sauvegarder et quitter")
    print("Choisissez uniquement un nombre entre 1 et 7.")
    print()


def main():
    print("Projet educatif : gestionnaire de mots de passe local.")
    print("Le fichier du coffre reste chiffre sur le disque.")

    coffre, fernet, entrees = ouvrir_ou_creer_coffre()

    while True:
        afficher_menu()
        choix = lire_texte("Votre choix (1 a 7) : ")

        if choix == "1":
            ajouter_entree(entrees)
            sauvegarder_coffre(coffre, fernet, entrees)
            pause()
        elif choix == "2":
            rechercher_entree(entrees)
            pause()
        elif choix == "3":
            lister_services(entrees)
            pause()
        elif choix == "4":
            if modifier_entree(entrees):
                sauvegarder_coffre(coffre, fernet, entrees)
            pause()
        elif choix == "5":
            if supprimer_entree(entrees):
                sauvegarder_coffre(coffre, fernet, entrees)
            pause()
        elif choix == "6":
            outil_generation()
            pause()
        elif choix == "7":
            sauvegarder_coffre(coffre, fernet, entrees)
            print("\nCoffre sauvegarde. A bientot.")
            break
        else:
            print("Choix invalide. Entrez uniquement un nombre entre 1 et 7.")
            pause()


if __name__ == "__main__":
    main()

File: gestionnaire-mdp.js
(function () {
    var terminal = document.getElementById("gestionnaire-mdp");

    if (!terminal) {
        return;
    }

    var screen = document.getElementById("password-screen");
    var form = document.getElementById("password-form");
    var input = document.getElementById("password-input");
    var toggleButton = document.getElementById("password-toggle");
    var resetButton = document.getElementById("password-reset");
    var demoButton = document.getElementById("password-demo");

    var coffre;
    var etape;
    var motDePasseMaitre;
    var entreeTemporaire;
    var resultatsRecherche;
    var indexSelectionne;
    var champSecret;
    var secretVisible;

    function ajouterLigne(texte, classe) {
        var ligne = document.createElement("p");
        ligne.className = "money-line " + (classe || "normal");
        ligne.textContent = texte;
        screen.appendChild(ligne);
        screen.scrollTop = screen.scrollHeight;
    }

    function afficherTitre(titre) {
        ajouterLigne("");
        ajouterLigne("=".repeat(64), "muted");
        ajouterLigne(titre, "title");
        ajouterLigne("=".repeat(64), "muted");
    }

    function poserQuestion(texte, secret) {
        ajouterLigne(texte, "prompt");
        input.placeholder = texte;
        definirChampSecret(secret);
        input.focus({ preventScroll: true });
    }

    function definirChampSecret(secret) {
        champSecret = !!secret;
        secretVisible = false;
        input.type = champSecret ? "password" : "text";

        if (toggleButton) {
            toggleButton.hidden = !champSecret;
            toggleButton.textContent = "Afficher";
            toggleButton.setAttribute("aria-pressed", "false");
        }
    }

    function texteOuiNon(texte) {
        var valeur = texte.trim().toLowerCase();
        if (valeur === "o" || valeur === "oui") {
            return true;
        }
        if (valeur === "n" || valeur === "non") {
            return false;
        }
        return null;
    }

    function nombreAleatoire(maximum) {
        var tableau;

        if (window.crypto && window.crypto.getRandomValues) {
            tableau = new Uint32Array(1);
            window.crypto.getRandomValues(tableau);
            return tableau[0] % maximum;
        }

        return Math.floor(Math.random() * maximum);
    }

    function genererMotDePasse(longueur) {
        var minuscules = "abcdefghijklmnopqrstuvwxyz";
        var majuscules = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
        var chiffres = "0123456789";
        var speciaux = "!@#$%&*()-_=+?";
        var alphabet = minuscules + majuscules + chiffres + speciaux;
        var caracteres = [
            minuscules[nombreAleatoire(minuscules.length)],
            majuscules[nombreAleatoire(majuscules.length)],
            chiffres[nombreAleatoire(chiffres.length)],
            speciaux[nombreAleatoire(speciaux.length)]
        ];
        var i;
        var j;
        var temporaire;

        longueur = Math.max(longueur || 16, 8);

        while (caracteres.length < longueur) {
            caracteres.push(alphabet[nombreAleatoire(alphabet.length)]);
        }

        for (i = caracteres.length - 1; i > 0; i -= 1) {
            j = nombreAleatoire(i + 1);
            temporaire = caracteres[i];
            caracteres[i] = caracteres[j];
            caracteres[j] = temporaire;
        }

        return caracteres.join("");
    }

    function afficherMenu() {
        afficherTitre("MENU DU COFFRE-FORT");
        ajouterLigne("1. Ajouter un mot de passe");
        ajouterLigne("2. Rechercher une entree");
        ajouterLigne("3. Lister les services");
        ajouterLigne("4. Modifier une entree");
        ajouterLigne("5. Supprimer une entree");
        ajouterLigne("6. Generer un mot de passe fort");
        ajouterLigne("7. Sauvegarder et quitter");
        ajouterLigne("Choisissez uniquement un nombre entre 1 et 7.", "muted");
        etape = "menu";
        poserQuestion("Votre choix (1 a 7) :");
    }

    function afficherEntree(entree, afficherMotDePasse) {
        ajouterLigne("Service     : " + entree.service);
        ajouterLigne("Identifiant : " + entree.identifiant);
        ajouterLigne("Note        : " + (entree.note || "Aucune"));
        ajouterLigne("Mot de passe: " + (afficherMotDePasse ? entree.motDePasse : "********"));
    }

    function rechercher(terme) {
        var valeur = terme.trim().toLowerCase();
        var resultats = [];
        var i;
        var entree;

        for (i = 0; i < coffre.length; i += 1) {
            entree = coffre[i];
            if (
                entree.service.toLowerCase().indexOf(valeur) !== -1 ||
                entree.identifiant.toLowerCase().indexOf(valeur) !== -1
            ) {
                resultats.push(i);
            }
        }

        return resultats;
    }

    function demanderSelection(resultats, prochaineEtape) {
        var i;
        var entree;

        if (resultats.length === 0) {
            ajouterLigne("Aucun resultat trouve.", "error");
            afficherMenu();
            return;
        }

        resultatsRecherche = resultats;
        ajouterLigne("");
        ajouterLigne("Resultats :", "title");

        for (i = 0; i < resultats.length; i += 1) {
            entree = coffre[resultats[i]];
            ajouterLigne((i + 1) + ". " + entree.service + " - " + entree.identifiant);
        }

        etape = prochaineEtape;
        poserQuestion("Numero de l'entree a selectionner ou n pour revenir :");
    }

    function lireSelection(reponse) {
        var choix = reponse.trim().toLowerCase();
        var numero;

        if (choix === "n" || choix === "non") {
            afficherMenu();
            return null;
        }

        if (!/^\d+$/.test(choix)) {
            ajouterLigne("Choix invalide.", "error");
            poserQuestion("Numero de l'entree a selectionner ou n pour revenir :");
            return null;
        }

        numero = Number(choix);

        if (numero < 1 || numero > resultatsRecherche.length) {
            ajouterLigne("Numero inexistant.", "error");
            poserQuestion("Numero de l'entree a selectionner ou n pour revenir :");
            return null;
        }

        return resultatsRecherche[numero - 1];
    }

    function demarrer() {
        screen.textContent = "";
        input.disabled = false;
        input.value = "";
        definirChampSecret(false);
        coffre = [];
        motDePasseMaitre = "";
        entreeTemporaire = null;
        resultatsRecherche = [];
        indexSelectionne = null;
        etape = "creation_mdp";

        afficherTitre("GESTIONNAIRE DE MOTS DE PASSE");
        ajouterLigne("Demo web du projet Python.");
        ajouterLigne("La vraie version cree un fichier coffre_mdp.json chiffre sur le disque.");
        ajouterLigne("Ici, les donnees restent seulement dans cette page.");
        poserQuestion("Creez un mot de passe maitre :", true);
    }

    function chargerExemple() {
        screen.textContent = "";
        coffre = [
            {
                service: "Gmail",
                identifiant: "samy@gmail.com",
                motDePasse: "Gm!82pQa#2026",
                note: "Compte personnel"
            },
            {
                service: "GitHub",
                identifiant: "AdellouchSamy",
                motDePasse: "Gh#47LmP!91",
                note: "Depot des projets Python"
            }
        ];
        input.disabled = false;
        definirChampSecret(false);
        motDePasseMaitre = "demo";
        afficherTitre("COFFRE EXEMPLE CHARGE");
        ajouterLigne("Deux entrees de demonstration ont ete ajoutees.", "success");
        afficherMenu();
    }

    function traiterMenu(reponse) {
        var choix = reponse.trim();

        if (choix === "1") {
            entreeTemporaire = {};
            etape = "ajout_service";
            afficherTitre("AJOUTER UN MOT DE PASSE");
            poserQuestion("Service ou site :");
        } else if (choix === "2") {
            if (coffre.length === 0) {
                ajouterLigne("Le coffre est vide.", "error");
                afficherMenu();
                return;
            }
            etape = "recherche_terme";
            afficherTitre("RECHERCHER UNE ENTREE");
            poserQuestion("Recherche par service ou identifiant :");
        } else if (choix === "3") {
            afficherTitre("LISTE DES SERVICES");
            if (coffre.length === 0) {
                ajouterLigne("Le coffre est vide.");
            } else {
                coffre.forEach(function (entree, index) {
                    ajouterLigne((index + 1) + ". " + entree.service + " - " + entree.identifiant);
                });
            }
            afficherMenu();
        } else if (choix === "4") {
            if (coffre.length === 0) {
                ajouterLigne("Le coffre est vide.", "error");
                afficherMenu();
                return;
            }
            etape = "modifier_terme";
            afficherTitre("MODIFIER UNE ENTREE");
            poserQuestion("Entree a modifier :");
        } else if (choix === "5") {
            if (coffre.length === 0) {
                ajouterLigne("Le coffre est vide.", "error");
                afficherMenu();
                return;
            }
            etape = "supprimer_terme";
            afficherTitre("SUPPRIMER UNE ENTREE");
            poserQuestion("Entree a supprimer :");
        } else if (choix === "6") {
            etape = "generer_longueur";
            afficherTitre("GENERATEUR DE MOTS DE PASSE");
            poserQuestion("Longueur souhaitee (16 par defaut) :");
        } else if (choix === "7") {
            afficherTitre("FIN DU PROGRAMME");
            ajouterLigne("Coffre sauvegarde dans la version Python.", "success");
            ajouterLigne("Merci d'avoir utilise le gestionnaire de mots de passe.");
            input.disabled = true;
        } else {
            ajouterLigne("Choix invalide. Entrez uniquement un nombre entre 1 et 7.", "error");
            poserQuestion("Votre choix (1 a 7) :");
        }
    }

    function traiterAjout(reponse) {
        var choix;
        var longueur;

        if (etape === "ajout_service") {
            if (reponse.trim() === "") {
                ajouterLigne("Le service ne peut pas etre vide.", "error");
                poserQuestion("Service ou site :");
                return;
            }
            entreeTemporaire.service = reponse.trim();
            etape = "ajout_identifiant";
            poserQuestion("Identifiant ou email :");
        } else if (etape === "ajout_identifiant") {
            if (reponse.trim() === "") {
                ajouterLigne("L'identifiant ne peut pas etre vide.", "error");
                poserQuestion("Identifiant ou email :");
                return;
            }
            entreeTemporaire.identifiant = reponse.trim();
            etape = "ajout_generation";
            poserQuestion("Generer un mot de passe fort ? (o/n)");
        } else if (etape === "ajout_generation") {
            choix = texteOuiNon(reponse);
            if (choix === null) {
                ajouterLigne("Tapez o ou n.", "error");
                poserQuestion("Generer un mot de passe fort ? (o/n)");
                return;
            }
            if (choix) {
                entreeTemporaire.motDePasse = genererMotDePasse(16);
                ajouterLigne("Mot de passe genere : " + entreeTemporaire.motDePasse, "success");
                etape = "ajout_note";
                poserQuestion("Note facultative :");
            } else {
                etape = "ajout_mdp";
                poserQuestion("Mot de passe a enregistrer :", true);
            }
        } else if (etape === "ajout_mdp") {
            if (reponse.trim() === "") {
                ajouterLigne("Le mot de passe ne peut pas etre vide.", "error");
                poserQuestion("Mot de passe a enregistrer :", true);
                return;
            }
            entreeTemporaire.motDePasse = reponse;
            etape = "ajout_note";
            poserQuestion("Note facultative :");
        } else if (etape === "ajout_note") {
            entreeTemporaire.note = reponse.trim();
            coffre.push(entreeTemporaire);
            entreeTemporaire = null;
            ajouterLigne("Entree ajoutee au coffre.", "success");
            afficherMenu();
        } else if (etape === "generer_longueur") {
            longueur = 16;
            if (/^\d+$/.test(reponse.trim())) {
                longueur = Number(reponse.trim());
            }
            ajouterLigne("Mot de passe genere : " + genererMotDePasse(longueur), "success");
            afficherMenu();
        }
    }

    function traiterRecherche(reponse) {
        var index;

        if (etape === "recherche_terme") {
            demanderSelection(rechercher(reponse), "recherche_selection");
        } else if (etape === "recherche_selection") {
            index = lireSelection(reponse);
            if (index === null) {
                return;
            }
            afficherTitre("ENTREE TROUVEE");
            afficherEntree(coffre[index], true);
            afficherMenu();
        }
    }

    function traiterModification(reponse) {
        var index;

        if (etape === "modifier_terme") {
            demanderSelection(rechercher(reponse), "modifier_selection");
        } else if (etape === "modifier_selection") {
            index = lireSelection(reponse);
            if (index === null) {
                return;
            }
            indexSelectionne = index;
            etape = "modifier_mdp";
            ajouterLigne("Entree selectionnee : " + coffre[index].service + " - " + coffre[index].identifiant);
            poserQuestion("Nouveau mot de passe, AUTO pour generer, ou vide pour conserver :", false);
        } else if (etape === "modifier_mdp") {
            if (reponse.trim().toUpperCase() === "AUTO") {
                coffre[indexSelectionne].motDePasse = genererMotDePasse(16);
                ajouterLigne("Nouveau mot de passe genere : " + coffre[indexSelectionne].motDePasse, "success");
            } else if (reponse.trim() !== "") {
                coffre[indexSelectionne].motDePasse = reponse;
                ajouterLigne("Mot de passe modifie.", "success");
            } else {
                ajouterLigne("Mot de passe conserve.");
            }
            etape = "modifier_note";
            poserQuestion("Nouvelle note ou vide pour conserver :");
        } else if (etape === "modifier_note") {
            if (reponse.trim() !== "") {
                coffre[indexSelectionne].note = reponse.trim();
                ajouterLigne("Note modifiee.", "success");
            }
            indexSelectionne = null;
            afficherMenu();
        }
    }

    function traiterSuppression(reponse) {
        var index;
        var choix;

        if (etape === "supprimer_terme") {
            demanderSelection(rechercher(reponse), "supprimer_selection");
        } else if (etape === "supprimer_selection") {
            index = lireSelection(reponse);
            if (index === null) {
                return;
            }
            indexSelectionne = index;
            ajouterLigne("Selection : " + coffre[index].service + " - " + coffre[index].identifiant);
            etape = "supprimer_confirmation";
            poserQuestion("Confirmer la suppression ? (o/n)");
        } else if (etape === "supprimer_confirmation") {
            choix = texteOuiNon(reponse);
            if (choix === null) {
                ajouterLigne("Tapez o ou n.", "error");
                poserQuestion("Confirmer la suppression ? (o/n)");
                return;
            }
            if (choix) {
                coffre.splice(indexSelectionne, 1);
                ajouterLigne("Entree supprimee.", "success");
            } else {
                ajouterLigne("Suppression annulee.");
            }
            indexSelectionne = null;
            afficherMenu();
        }
    }

    form.addEventListener("submit", function (event) {
        var reponse;
        var affichage;

        event.preventDefault();

        if (input.disabled) {
            return;
        }

        reponse = input.value;
        affichage = champSecret && reponse !== "" ? "********" : reponse;
        ajouterLigne("> " + affichage, "answer");
        input.value = "";
        definirChampSecret(false);

        if (etape === "creation_mdp") {
            if (reponse.length < 8) {
                ajouterLigne("Le mot de passe maitre doit contenir au moins 8 caracteres.", "error");
                poserQuestion("Creez un mot de passe maitre :", true);
                return;
            }
            motDePasseMaitre = reponse;
            etape = "confirmation_mdp";
            poserQuestion("Confirmez le mot de passe maitre :", true);
        } else if (etape === "confirmation_mdp") {
            if (reponse !== motDePasseMaitre) {
                ajouterLigne("Les deux mots de passe ne correspondent pas.", "error");
                etape = "creation_mdp";
                poserQuestion("Creez un mot de passe maitre :", true);
                return;
            }
            ajouterLigne("Coffre cree avec succes.", "success");
            afficherMenu();
        } else if (etape === "menu") {
            traiterMenu(reponse);
        } else if (
            etape === "ajout_service" ||
            etape === "ajout_identifiant" ||
            etape === "ajout_generation" ||
            etape === "ajout_mdp" ||
            etape === "ajout_note" ||
            etape === "generer_longueur"
        ) {
            traiterAjout(reponse);
        } else if (etape === "recherche_terme" || etape === "recherche_selection") {
            traiterRecherche(reponse);
        } else if (
            etape === "modifier_terme" ||
            etape === "modifier_selection" ||
            etape === "modifier_mdp" ||
            etape === "modifier_note"
        ) {
            traiterModification(reponse);
        } else if (
            etape === "supprimer_terme" ||
            etape === "supprimer_selection" ||
            etape === "supprimer_confirmation"
        ) {
            traiterSuppression(reponse);
        }
    });

    if (toggleButton) {
        toggleButton.addEventListener("click", function () {
            if (!champSecret) {
                return;
            }

            secretVisible = !secretVisible;
            input.type = secretVisible ? "text" : "password";
            toggleButton.textContent = secretVisible ? "Masquer" : "Afficher";
            toggleButton.setAttribute("aria-pressed", secretVisible ? "true" : "false");
            input.focus({ preventScroll: true });
        });
    }

    resetButton.addEventListener("click", demarrer);
    demoButton.addEventListener("click", chargerExemple);

    demarrer();
}());
